bach/bach/operations/cut.py

```diff
@@ -222,5 +224,117 @@ def _calculate_bucket_ranges(self, bucket_properties_df: 'DataFrame') -> 'Series
+                expression=Expression.construct(range_stmt, df.all_series[self.series.name]),
+                name=self.RANGE_SERIES_NAME,
+            )
+
+            df = df.merge(quantile_ranges, how='left', on=self.RANGE_SERIES_NAME)
+
+        new_index = {self.series.name: df.all_series[self.series.name]}
+        return cast(SeriesFloat64, df.all_series[self.RANGE_SERIES_NAME].copy_override(index=new_index))
+
+    def _get_quantile_ranges(self) -> 'Series':
+        """
+        Calculates the corresponding ranges per each quantile bucket.
+
+        The following series are calculated:
+        * lower_bound: for calculating the lower bound of each range it is required to calculate
+        all requested quantiles from the series. The lowest bound is adjusted since it might be
+        included in the dataset.
+        * upper_bound: is the lower bound from the next range that follows the current one.
+         If current lower bound is the result of the largest quantile, the upper bound will be null.
+        * range: interval containing the calculated upper and lower bounds per bucket.
+        If the upper bound is null, the range will be null.
+
+        Returns a series containing the range of each bucket
+        """
+        q_result = self.series.quantile(q=self.quantiles).copy_override(name='q_result')
+        min_q_result = q_result.min()
+
+        quantile_ranges_df = q_result.to_frame()
+
+        # some quantiles might have the same result, we need to avoid having overlapped ranges
+        quantile_ranges_df.drop_duplicates(inplace=True, ignore_index=True)
+
+        # lowest calculated quantile might be also be in the dataset, therefore
+        # we need to extend the lowest bound
+        quantile_ranges_df['lower_bound'] = quantile_ranges_df.all_series['q_result'].copy_override(
+            expression=Expression.construct(
+                f'case when {{}} = {{}} then {{}} - {_RANGE_ADJUSTMENT} else {{}} end',
+                Series.as_independent_subquery(min_q_result),
+                *[quantile_ranges_df.all_series['q_result']] * 3
+            )
+        )
+
+        # should call "series.lag" instead but just need sorting in the window function
+        quantile_ranges_df['upper_bound'] = quantile_ranges_df.all_series['lower_bound'].copy_override(
+            expression=Expression.construct(
+                f'lag({{}}, 1, NULL) over (order by {{}} desc)',
+                quantile_ranges_df.all_series['lower_bound'],
+                quantile_ranges_df.all_series['lower_bound'],
+            ),
+            name='upper_bound'
+        )
+
+        # must not include lower bound, since a calculated quantile might also be in the dataset
+        # so the value can generate duplicates
+        bound = "'(]'"
+        range_stmt = (
+            f'case when {{}} is not null\n'
+            f'then numrange(cast({{}} as numeric), cast({{}} as numeric), {bound}) end'
+        )
```

Comment:
```suggestion
        )
        quantile_ranges_df.materialize(inplace=True)
```

Checked this on one test, saved about 600 bytes of generated sql. But more importantly made the sql much more readable